bplustree_test.go

```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
+		t.FailNow()
+	}
+
+	if len(result) != 5 {
+		t.Errorf("Expected to find 5 results, got: %d", len(result))
+		t.FailNow()
+	}
+
+	// Note that our values are +10 of the key.
+	startKey := makeTestingKeyStr(strHash(testingKey) - 2)
+	expectedVal := int(strHash(startKey) + 10)
+	for i := 0; i < len(result); i++ {
+		//fmt.Println(reflect.TypeOf(result[i]))
+		actualVal := result[i].(*TestingElemStr).val
+		if actualVal != expectedVal {
+			t.Errorf("At index %d: expected: %d, got %v", i, expectedVal, actualVal)
+			t.FailNow()
+		}
+		// and increment by 1.
+		expectedVal++
+	}
+}
+
+func TestRemoveStr1(t *testing.T) {
+	fmt.Println("Starting remove in order tests")
+	for i := 0; i < numElems; i++ {
+		key := makeTestingKeyStr(i)
+		fmt.Printf("removing key: %v", key)
+		err := newtree.Remove(key)
+		if err != nil {
+			t.Errorf("Failed to remove key: %v got %v", key, err)
+			t.FailNow()
+		}
+	}
+}
+
+func TestRemoveStr2(t *testing.T) {
+	TestInsertStr(t)
+	fmt.Println("Starting remove in reverse order tests")
+	for i := numElems - 1; i >= 0; i-- {
+		key := makeTestingKeyStr(i)
+		fmt.Printf("removing key: %v", key)
+		err := newtree.Remove(key)
+		if err != nil {
+			t.Errorf("Failed to remove key: %v got %v", key, err)
+			t.FailNow()
+		}
+	}
+}
+
+func TestRemoveStr3(t *testing.T) {
+	TestInsertStr(t)
+	fmt.Println("Starting remove in weird order test")
+
+	removed := make(map[int]bool)
+	for i := 0; i < numElems/3; i++ {
+		k1 := i
+		key := makeTestingKeyStr(k1)
+		fmt.Printf("removing key: %v", key)
+		err := newtree.Remove(key)
+		if err != nil {
+			t.Errorf("Failed to remove key: %v got %v", key, err)
+			t.FailNow()
+		}
+		removed[k1] = true
+
+		k2 := numElems/3 + i
+		key = makeTestingKeyStr(k2)
+		fmt.Printf("removing key: %v", key)
+		err = newtree.Remove(key)
+		if err != nil {
+			t.Errorf("Failed to remove key: %v got %v", key, err)
+			t.FailNow()
+		}
+		removed[k2] = true
+
+		k3 := numElems*2/3 + i
+		key = makeTestingKeyStr(k3)
+		fmt.Printf("removing key: %v", key)
+		err = newtree.Remove(key)
+		if err != nil {
+			t.Errorf("Failed to remove key: %v got %v", key, err)
+			t.FailNow()
+		}
+		removed[k3] = true
+	}
+
+	for i := 0; i < numElems; i++ {
+		if !removed[i] {
+			newtree.Remove(makeTestingKeyStr(i))
+		}
+	}
+}
+
+func RemoveRoutineStr(t *testing.T, k int, errors []error, index int, wg *sync.WaitGroup) {
+	defer wg.Done()
+	t.Logf("Removing: %v", k)
+	err := newtree.Remove(makeTestingKeyStr(k))
+	errors[index] = err
+	t.Logf("++++++++++++++++++++++++++++++++")
+}
+
+func TestRemoveStr4(t *testing.T) {
+	TestInsertStr(t)
+	fmt.Println("Starting remove in parallel test")
+	errors := make([]error, numElems)
+	var wg sync.WaitGroup
+
+	for i := 0; i < numElems; i++ {
+		wg.Add(1)
+		go RemoveRoutineStr(t, i, errors, i, &wg)
+	}
+
+	wg.Wait()
+	for i := 0; i < numElems; i++ {
+		if errors[i] != nil {
+			t.Errorf("Failed to insert: %d: %v", i, errors[i])
+			t.FailNow()
+		}
+	}
+	t.Logf("\nFinal tree:\n")
```

Comment:
you are mostly removing all elements. how about adding cases to make sure rebalancing happens properly on removals that cause numelements to be maxsize/2?
I feel that the edge cases are not being test here.

Reply:
They are guaranteed to happen.. there's no way to validate that redistribute/merge happen programatically.. but going in the deterministic way, like in order or in reverse order, we are guaranteed to have both these occur upon deletion of of consecutive 'degree' # of elements. in order and reverse order also guarantee that we are covering all edge cases of prev/next etc. being null in certain conditions. And also I have things being removed from middle also every now and then..